lessons/advanced/protocols.md

```diff
@@ -0,0 +1,78 @@
+---
+layout: page
+title: Protocols
+category: advanced
+order: 12
+lang: en
+---
+
+In this lesson we are going to look at Protocols, and how to leverage them in Elixir.
+
+{% include toc.html %}
+
+## What Are Protocols
+Protocols are the chosen method for polymorphism in Elixir.
+One of the primary pains of Erlang is that extending a given API for a newly defined type is quite difficult.
+What Elixir does to avoid this pain is dispatch function calls dynamically based on the type of a value provided.
+This typically is not an issue in Object Oriented languages, but the workaround is necessary in our environment here.
+If a data type implements a given protocol, it will have it available to dispatch to.
+Let's demonstrate this with a simple example looking at `to_string/1`.
+
+```elixir
+iex> to_string(5)
+"5"
+iex> to_string(12.4)
+"12.4"
+iex> to_string("foo")
+"foo"
+```
+
+As you can see, we have called this on multiple types.
+We've demonstrated that the function works on integers, floats, and strings as well.
+If we try to do it on a tuple (or anything else that has not implemented it), we get a different result:
+
+```elixir
+to_string({:foo})
+** (Protocol.UndefinedError) protocol String.Chars not implemented for {:foo}
+    (elixir) lib/string/chars.ex:3: String.Chars.impl_for!/1
+    (elixir) lib/string/chars.ex:17: String.Chars.to_string/1
+```
+
+As you can see, we get a protocol error.
+This is because there is no implementation for the Protocol in the tuples.
```

Comment:
As you can see we get a protocol error as there is no implementation for tuples.  In the next section we'll implement the `String.Chars` protocol for tuples.

## Implementing a protocol

We saw that `to_string/1` has not yet been implemented for tuples so let's add it.  To create an implementation we'll use `defimpl` with our protocol, and provide the `:for` option and our type.  Let's take a look at how it might look:

```elixir
defimpl String.Chars, for: Tuple do
  def to_string(tuple) do
    insides = 
      tuple
      |> Tuple.to_list
      |> Enum.map(&Kernel.to_string/1)
      |> Enum.join(", ")

    "{#{insides}}"
  end
end
```

If we copy this into IEx we should be now be able to call `to_string/1` on a tuple without getting an error:

```elixir
iex> to_string({3.14, "apple", :pie})
"{3.14, apple, pie}"
```